### PR TITLE
feat(ci): validate state version with cached state

### LINKED ISSUE #2256: Add the state version to the Google Cloud cached state disk names
## Motivation

When we're looking at the disk names for our CI cached state, it's hard to tell what Zebra state version they contain.

### Designs

https://github.com/ZcashFoundation/zebra/blob/2e0318878a1b95d72d762245816362a7d812ef5c/zebra-state/src/constants.rs#L28-L29

## Solution

- [ ] move the state version number to a file, and use `include!` to import that number into `constants.rs`
- [x] put the state version number in the disk names for the CI cached state

### Alternatives

We could use `grep` to find the value of the state version constant. Or we could write a CLI tool that outputs this information.

If we miss a state bump, our cached state tests will timeout and let us know we need an update.

## Related Work

Originally discussed in https://github.com/ZcashFoundation/zebra/pull/2247#discussion_r645257146

### LINKED ISSUE #3881: Old PRs can add a new "latest" cached state w/ the old state version
## Motivation

We don't check the state version when updating the cached state, so old PRs can add a new "latest" cached state with the old state version. This won't break CI, but it adds 2.5 hours to the mergify queue.

We can check cached state version vs the version in the current branch being tested. Doing this requires making the cached state version easier to find in CI. One idea is adding the state version as disk image metadata or in the name of the disk, and then doing some Rust test / grep check to compare the latest cached state disk version with the in-branch version, and only using the cached state if they are the same version. 

If the cached state version is higher than the branch version, we can not run the test, failing the test until the branch is rebased onto #main.

If the cached state version is lower than the branch version, we can run the test from an empty disk? Requires a full sync to the stop point.

Either way, the goal is that 'old' state versions (older than the version of state on #main) should not be created and pushed to gcloud, where they can be used by other tests.

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

We don't check the state version when updating the cached state, so old PRs can add a new "latest" cached state with the old state version. This won't break CI, but it adds 2.5 hours to the mergify queue.

We can check cached state version vs the version in the current branch being tested. Doing this requires making the cached state version easier to find in CI. One idea is adding the state version as disk image metadata or in the name of the disk, and then doing some Rust test / grep check to compare the latest cached state disk version with the in-branch version, and only using the cached state if they are the same version.

If the cached state version is higher than the branch version, we can not run the test, failing the test until the branch is rebased onto `main`.

Either way, the goal is that 'old' state versions (older than the version of state on #main) should not be created and pushed to gcloud, where they can be used by other tests.

Fixes #3881
Depends-On: #4073

## Solution
- Before building or running other tests, confirm the local (branch) state version against the cached state version

## Review
@dconnolly can review this

## Follow Up Work

- Decide if there's a different behavior to be done when the `cached state version` is lower than the `branch version`
- Creating images when pushing to main to reduce cross-PR contamination

### CODE REVIEW

File: .github/workflows/test-full-sync.yml
Comment (teor2345) [inline .github/workflows/test-full-sync.yml]: Typo?

```suggestion
      # or at least that the local state version is higher than the available cached state version.
```

File: .github/workflows/test-full-sync.yml
Comment (teor2345) [inline .github/workflows/test-full-sync.yml]: This comment does not apply to the full sync test, because it always starts from genesis.

File: .github/workflows/test-full-sync.yml
Comment (teor2345) [inline .github/workflows/test-full-sync.yml]: Possible Bug: It looks like this check is skipped if the state version changes between the PR and the main branch. But it seems like we always want to check the state version here.

(If that's the case, can you also delete the corresponding comment?)

File: .github/workflows/test.yml
Comment (teor2345) [inline .github/workflows/test.yml]: Note: this test could also fail if the state versions are incompatible

```suggestion
      # we want to run with the old cached state, so this test or other tests fail.
```

File: .github/workflows/test.yml
Comment (teor2345) [inline .github/workflows/test.yml]:
```suggestion
      # or at least that the local state version is higher than the available cached state version.
```

File: .github/workflows/test.yml
Comment (teor2345) [inline .github/workflows/test.yml]: Question: I'm not sure I understand why we are skipping this check in this test. What would happen if we always did the check?

**Review by teor2345 — COMMENTED** [review]
I'm not sure I understand exactly what is going on in these tests, happy to catch up for a video call if that would help.

File: .github/workflows/test-full-sync.yml
Comment (teor2345) [inline .github/workflows/test-full-sync.yml]: Question: do we also want to run this check when the workflow file changes?

File: .github/workflows/test.yml
Comment (teor2345) [inline .github/workflows/test.yml]: Question: do we also want to regenerate disks when the workflow file changes?

File: .github/workflows/test-full-sync.yml
Comment (gustavovalverde) [reply .github/workflows/test-full-sync.yml]: Fixed by: https://github.com/ZcashFoundation/zebra/pull/4074/commits/8f24f0f0dc62c307adc43d74e283e27950f9abe4

File: .github/workflows/test-full-sync.yml
Comment (gustavovalverde) [reply .github/workflows/test-full-sync.yml]: Fixed by: https://github.com/ZcashFoundation/zebra/pull/4074/commits/8f24f0f0dc62c307adc43d74e283e27950f9abe4

File: .github/workflows/test.yml
Comment (gustavovalverde) [reply .github/workflows/test.yml]: Clarified with: https://github.com/ZcashFoundation/zebra/pull/4074/commits/8f24f0f0dc62c307adc43d74e283e27950f9abe4

File: .github/workflows/test.yml
Comment (gustavovalverde) [reply .github/workflows/test.yml]: Fixed by: https://github.com/ZcashFoundation/zebra/pull/4074/commits/8f24f0f0dc62c307adc43d74e283e27950f9abe4

File: .github/workflows/test-full-sync.yml
Comment (gustavovalverde) [reply .github/workflows/test-full-sync.yml]: Fixed by: https://github.com/ZcashFoundation/zebra/pull/4074/commits/8f24f0f0dc62c307adc43d74e283e27950f9abe4

File: .github/workflows/test.yml
Comment (gustavovalverde) [reply .github/workflows/test.yml]: Fixed by: https://github.com/ZcashFoundation/zebra/pull/4074/commits/8f24f0f0dc62c307adc43d74e283e27950f9abe4

**Comment by gustavovalverde** [comment]
I've unliked #2256 and added it to #4073

File: .github/workflows/test.yml
Comment (gustavovalverde) [reply .github/workflows/test.yml]: This would add a few hours to most PRs, if we'd like to do this for specific files it would be better to add those to this step: https://github.com/ZcashFoundation/zebra/pull/4074/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R77

Comment (teor2345) [reply .github/workflows/test.yml]: It seems like the long-term solution is to split this job and its dependencies into their own workflow file.

Do we have a ticket for that?

Comment (gustavovalverde) [reply .github/workflows/test.yml]: I'll create the ticket for `reusable workflows` which, by the way, is a considerable change to be done before working in pipeline documentation. I think it's even better to prioritize that somehow.

**Review by teor2345 — COMMENTED** [review]
I'm happy to approve this PR, but can we run a full set of CI tests first?

It might be best to wait until #4073 merges, then run the tests.

**Comment by gustavovalverde** [comment]
This is interesting, it seems one of the PRs is being worked with a database upgrade, but that PR hasn't been merged yet and the cached state has been pushed to GCP, making my PR fail https://github.com/ZcashFoundation/zebra/runs/5992738082?check_suite_focus=true#step:6:79

The scenario is the following
- My PR has a lower version than the `cached` version
- The local version in `main` is lower than the `cached` version
- A PR updated the `cached` version, but it hasn't been merged into `main`

I'm still not sure on how to handle this scenario, but so far is clear that:
- If we update `constants.rs` in a PR, we shouldn't push the disk image
- At the same time, we don't want every commit from this PR upgrading the DB to be syncing from genesis each time

**Comment by teor2345** [comment]
> This is interesting, it seems one of the PRs is being worked with a database upgrade, but that PR hasn't been merged yet and the cached state has been pushed to GCP, making my PR fail 

The cached state upgrade rules in this PR only apply to the cached state version on the `main` branch, so we might need to make this change in this PR:
> > Creating images when pushing to main to reduce cross-PR contamination

Or we could handle `main` cached state differently from any other cached states. This seems like the smallest fix.

> * At the same time, we don't want every commit from this PR upgrading the DB to be syncing from genesis each time

If the PR has a higher state version than `main`, we could create a cached state for:
* only this PR, or
* only PRs that have the same state version

Then when a PR with a higher state version is merged to `main`, we could update the `main` cached state.

**Comment by gustavovalverde** [comment]
@teor2345 I've integrated the considerations from your last comment.

File: .github/workflows/test-full-sync.yml
Comment (teor2345) [inline .github/workflows/test-full-sync.yml]: Optional Lint: This should fix the shellcheck lint:

```suggestion
          LOCAL_STATE_VERSION=$(grep -oE "DATABASE_FORMAT_VERSION: .* [0-9]+" "$GITHUB_WORKSPACE/zebra-state/src/constants.rs" | grep -oE "[0-9]+" | tail -n1)
```

File: .github/workflows/test.yml
Comment (teor2345) [inline .github/workflows/test.yml]:
```suggestion
          LOCAL_STATE_VERSION=$(grep -oE "DATABASE_FORMAT_VERSION: .* [0-9]+" "$GITHUB_WORKSPACE/zebra-state/src/constants.rs" | grep -oE "[0-9]+" | tail -n1)
```

File: .github/workflows/test.yml
Comment (teor2345) [inline .github/workflows/test.yml]: Question: we're repeating this condition twice, is there a way to just depend on the condition from the previous step?

(It's ok if there isn't.)

File: .github/workflows/test.yml
Comment (teor2345) [inline .github/workflows/test.yml]: Optional Documentation: why is this condition different from the previous steps?

It is explained below, but it was not clear to me from this comment.

File: .github/workflows/test.yml
Comment (teor2345) [inline .github/workflows/test.yml:406]: Optional Lint: you could fix this shellcheck issue by splitting the subshell out, and assigning the result to its own variable.

This might also make the code easier to read.

File: .github/workflows/test.yml
Comment (teor2345) [inline .github/workflows/test.yml:408]:
```suggestion
          echo "CACHED_DISK_NAME=$GCP_STATE_DISK" >> "$GITHUB_ENV"
```

File: .github/workflows/test.yml
Comment (teor2345) [inline .github/workflows/test.yml:404]: Performance Question: Do we want to re-use any matching state version here, rather than strictly matching the branch name?

If we match the branch name, we will rebuild the state twice for each database format PR:
* in the original PR
* in the Mergify queue PR

**Review by teor2345 — COMMENTED** [review]
Looks good, but I have a bunch of questions about how it works in detail.

File: .github/workflows/test.yml
Comment (gustavovalverde) [reply .github/workflows/test.yml:408]: I'd have to do this globally, so I'd prefer to do this on another PR

File: .github/workflows/test.yml
Comment (gustavovalverde) [reply .github/workflows/test.yml]: Fixed by [imp(ci): do not repeat conditions if not needed](https://github.com/ZcashFoundation/zebra/pull/4074/commits/3372cd34c863dce304dcf29462dc34d8657e7db1)

File: .github/workflows/test.yml
Comment (teor2345) [reply .github/workflows/test.yml:408]: No worries, it's not a blocker at all.

Feel free to ignore all the shellcheck suggestions.

File: .github/workflows/test.yml
Comment (gustavovalverde) [reply .github/workflows/test.yml:406]: I'll do a full shellcheck fix PR, so this one doesn't get messy or implement a different approach from other files

File: .github/workflows/test.yml
Comment (gustavovalverde) [reply .github/workflows/test.yml]: Fixed by: [docs(ci): improve reasoning behind new disk selection conditions](https://github.com/ZcashFoundation/zebra/pull/4074/commits/3d7aec7512fce67adcbe97e12082a1847ee2098e)

File: .github/workflows/test.yml
Comment (gustavovalverde) [reply .github/workflows/test.yml:404]: Maybe this comments applies to the `regenerate-stateful-disks` job, from here on https://github.com/ZcashFoundation/zebra/blob/feat-state-versioning/.github/workflows/test.yml#L294

On a PR, so far, we've been re-building the state based on this condition:
`${{ steps.changed-files-specific.outputs.any_changed == 'true' || github.event.inputs.regenerate-disks == 'true'}}`

If the `constants.rs` is changed in a commit, the rebuild has been happening on each new commit on that specific PR and also in the Mergify queue.

> Do we want to re-use any matching state version here, rather than strictly matching the branch name?

@teor2345 I don't think I have enough knowledge on the behavior from using matching-version states available from other PRs, or if there's any edge-case that might cause syncing from genesis, or an unexpected failure. I took this approach to ensure there's no state contamination between PRs, but maybe it's not needed. I'll let you make that call and define the desired design (maybe in a ticket to get it done in another PR?)

Comment (teor2345) [reply .github/workflows/test.yml:404]: > If the `constants.rs` is changed in a commit, the rebuild has been happening on each new commit on that specific PR and also in the Mergify queue.

If we merge this PR, the performance of state rebuilds will be worse than the current CI behaviour. We will rebuild on every commit. But currently, we rebuild the cache once, then use it as long as the state version stays the same.

Checking the state version is a low priority, so I'm not sure if we should merge this PR. What do you think?

I would be happy to merge version checks that have similar performance.

> > Do we want to re-use any matching state version here, rather than strictly matching the branch name?
> 
> @teor2345 I don't think I have enough knowledge on the behavior from using matching-version states available from other PRs, or if there's any edge-case that might cause syncing from genesis, or an unexpected failure. I took this approach to ensure there's no state contamination between PRs, but maybe it's not needed.

The goal was to share cached state between PRs and commits, not to isolate the state from each commit. We need to do this for performance reasons - otherwise every `constants.rs` commit takes 3 hours.

I'm sorry for the confusion, hopefully we can be clearer next time. 😢 

> I'll let you make that call and define the desired design (maybe in a ticket to get it done in another PR?)

It seems like we have a few open low-priority DevOps PRs, and some of them will (temporarily) make CI performance worse.

But Zebra needs a cached state for the integration tests, and that is a high priority.

So I'm not sure what the best way forward is here.

I wonder if there is a quick fix that unblocks the high priority tasks, and lets us keep working on the low priority tasks when we have time? Or a quick change that avoids the performance issues?

Maybe we can catch up and talk through the details? Is the testing meeting tomorrow ok?

Comment (gustavovalverde) [reply .github/workflows/test.yml:404]: Yeah, let's have a time tomorrow in the testing meeting to go through this one.

Comment (gustavovalverde) [reply .github/workflows/test.yml:404]: Just to clear the path a bit before tomorrow's meeting:

This PR is not changing the condition on `when` the rebuild is happening, aside from this one:
https://github.com/ZcashFoundation/zebra/pull/4074/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R294

And it's basically creating a constraint to not rebuild on each push (which is happening right now, before this PR).

The conditions in `get-latest-disk-name` and `get-main-disk-name` are just for disk selection.

> But currently, we rebuild the cache once, then use it as long as the state version stays the same.

Based on the actual condition `if: ${{ steps.changed-files-specific.outputs.any_changed == 'true' || github.event.inputs.regenerate-disks == 'true' || github.event_name == 'push'}}` we're rebuilding on each push to  a PR (when `constants.rs` is changed) and rebuilding when pushing to `main` no matter the condition.

Comment (teor2345) [reply .github/workflows/test.yml:404]: Ok, well it seems like we should just merge this, and fix it up later?

I opened ticket #4153 for that fix, we can decide if anything depends on it at the testing meeting.

**Comment by teor2345** [comment]
Regenerate stateful disks failed with #3991:
> unexpected EOF

I'll restart the job.

File: .github/workflows/test.yml
Comment (teor2345) [reply .github/workflows/test.yml:404]: (I'll approve the PR, if you're happy with this solution, feel free to resolve this comment and queue it for merge.)

**Review by teor2345 — DISMISSED** [review]
Thanks for your patience with this!

**Comment by teor2345** [comment]
@mergifyio update

**Comment by gustavovalverde** [comment]
This can't be merged until we fix https://github.com/ZcashFoundation/zebra/issues/4155

**Comment by teor2345** [comment]
Please include the clean fix #4206 when next updating this branch.

**Comment by teor2345** [comment]
This isn't in the critical path for lightwalletd work, we can merge it after we get the lightwalletd integration tests working.

**Comment by gustavovalverde** [comment]
Obsoleted by #4271

**Pull request closed without merging**
